Pipeline Implementation

vaikas-google opened the pull request:
Fixes #1067 

## Proposed Changes

- Add messaging.knative.dev/pipelines for being able to define simple sequences

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Adds a pipeline (sequence of functions to execute in order) with an optional Reply to send results to.
```

akashrv commented:
/milestone v0.7.0

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_lifecycle.go:
```diff
@@ -0,0 +1,57 @@
+
+	// PipelineConditionChannelReady has status True when controller has successfully added a
+	// pipeline to the spec.channel resource.
+	PipelineConditionChannelReady duckv1alpha1.ConditionType = "ChannelReady"
```
Does this represent all Channels used in the pipeline? Also, the comment about `spec.channel` doesn't make sense to me.

vaikas-google replied:
Sorry, there's left over cruft, please just look at the pipeline_types.go file. Sorry, tried to direct attention to ONLY that file :)

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,96 @@
+
+	// ChannelStatuses is an array of corresponding Channel statuses.
+	// Matches the Spec.Steps array in the order.
+	ChannelStatuses []eventingv1alpha1.ChannelStatus
```
This will include stuff like the address of each intermediate Channel. I'm guessing we don't want that. How about

```golang
type PipelineChannelStatus struct {
  Channel v1.ObjectReference
  ReadyCondition duckv1alpha1.Condition
}
```

vaikas-google replied:
Great idea!

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,96 @@
+
+	// SubscriptionStatuses is an array of corresponding Subscription statuses.
+	// Matches the Spec.Steps array in the order.
+	SubscriptionStatuses []eventingv1alpha1.SubscriptionStatus
```
This also includes a lot of extra information. I would include just a reference to the Subscription and its Ready condition.

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,96 @@
+	// SubscriptionStatuses is an array of corresponding Subscription statuses.
+	// Matches the Spec.Steps array in the order.
+	SubscriptionStatuses []eventingv1alpha1.SubscriptionStatus
+
```
Should we also have the status of the final `spec.Subscriber` Subscription? Or is that implicitly included in the final `spec.steps`'s Subscription?

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,96 @@
+	// provided.
+	Steps []eventingv1alpha1.SubscriberSpec
+
+	// TODO: Specify the CRD Channel which should be used to create the underlying Channels.
```
This seems important. It is fine leaving it as a TODO for now, but I consider it a blocker for actually releasing Pipelines.

vaikas-google replied:
Yes, it has to be done before this goes in. Was napkinning what this would look like last night.

vaikas-google replied:
Good now?

vaikas-google replied:
@Harwayne

Harwayne replied:
LGTM

lionelvillard commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,96 @@
+
+	// Subscriber is the addressable that optionally receives replies from the last step in the pipeline.
+	// +optional
+	Subscriber *eventingv1alpha1.SubscriberSpec `json:"subscriber,omitempty"`
```
Why is this needed? Can it be the last `Steps`?

vaikas-google replied:
Good question. I got this type wrong. It should have been ReplyStrategy or something like that (it assumes currently only a channel, but there's an issue to relax this constraint to others as well). Basically for the last subscription (representing the last function in the pipeline) created by the pipeline controller, there needs to be a way to specify the Subscription.Spec.Reply. For example, if I have a pipeline with 2 functions and the last one produces the "final" output of this pipeline, I might want to wire that into the Broker for others to consume, or push it to some other channel if strictly operating in the messaging (Channel / Subscription bases).

lionelvillard commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,134 @@
+	Steps []eventingv1alpha1.SubscriberSpec `json:"steps"`
+
+	// ChannelTemplate specifies which Channel CRD to use
+	ChannelTemplate ChannelTemplateSpec `json:"channelTemplate"`
```
Don't we want each step to use an existing channel or create their own channel? I'm afraid this design implies that a channel is created per step.

vaikas-google replied:
Sorry, use which existing channel? If you have channels you're already created, seems like pipeline does not really add that much value since you can just create your subscriptions just as easily? I was reading through the initial proposal that we talked about a while ago, and there was no mention of reusing channels, rather than the pipeline would create the channels and subscriptions necessary to wire things together, so the user would simply give a list of functions that get executed in order.

lionelvillard replied:
yes make sense, now that I understand better what can be done with handlers...

lionelvillard commented:
@vaikas-google If you want I can take an initial stab at the reconciler. LMK.

mattmoor-sockpuppet commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,137 @@
+var _ apis.Validatable = (*Pipeline)(nil)
+var _ apis.Defaultable = (*Pipeline)(nil)
+
+// TOOD: make appropriate fields immutable.
```
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,138 @@
+var _ apis.Validatable = (*Pipeline)(nil)
+var _ apis.Defaultable = (*Pipeline)(nil)
+
+// TOOD: make appropriate fields immutable.
```
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet replied:
<!--typo-bot-->
```suggestion
// TODO: make appropriate fields immutable.
```

mattmoor-sockpuppet commented on config/300-pipeline.yaml:
```diff
@@ -0,0 +1,91 @@
+# Copyright 2018 The Knative Authors
```
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

mattmoor-sockpuppet replied:
<!--copyright-bot-->
```suggestion
# Copyright 2019 The Knative Authors
```

vaikas-google commented:
@Harwayne @lionelvillard 
Ok, I think it's ready for a proper review. PTAL. Thanks :)

lionelvillard commented on config/300-pipeline.yaml:
```diff
@@ -0,0 +1,91 @@
+                  type: string
+                  minLength: 1
+                spec:
+                  type: object
```
How does this get validated?

n3wscott replied:
you can't in the openapi doc. You could if you loaded the channel template crd. That can be done in the future.

vaikas-google replied:
Yeah, so it should fail now at CR creation time, so while we can't easily ( as Scott said) validate the target channel, at least when we try to create it will fail instead of previous provisioner model where the channel would get created and then fail at a later time. I'd like to not tackle this as part of this PR. One option would be to use dry run. Regardless, created this issue to track that, ok?
https://github.com/knative/eventing/issues/1349

lionelvillard commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,289 @@
+	}
+
+	// Convert the object into runtime.Object so we can grab the schema.ObjectKind and create the correct interface client
+	obj := p.Spec.ChannelTemplate.DeepCopyObject()
```
Why deep copy? It seems unnecessary

n3wscott replied:
this code is cloning the template.

vaikas-google replied:
not necessary, removed.

lionelvillard commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,289 @@
+		if err != nil {
+			// TODO: Send events here, or elsewhere?
+			//			r.Recorder.Eventf(p, corev1.EventTypeWarning, subscriptionCreateFailed, "Create Pipeline's subscription failed: %v", err)
+			return nil, err
```
The error should be logged

vaikas-google replied:
it gets logged if this method fails at the call site on L208, seemed redundant to log it twice?

vaikas-google replied:
added a log statement regardless.

Harwayne replied:
My thoughts have been going the other way. Rather than log at each function level. Attach the additional context at each function level, then the log only when it is no longer 'fatal' (for reconcilers, normally the `reconcile()` method, which recovers from the error).

So this function would be something like:

```golang
if err != nil {
    return nil, fmt.Errorf("failed to create subscription object: %v", err)
}
```

Then 208 (now 217) becomes something like:

```golang
if err != nil {
    return nil, fmt.Errorf("failed to reconcile Subscription Object for step %d: %v", step, err)
}
```

Which ends up getting logged exactly once in the reconcile() method (line 137).

lionelvillard commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,289 @@
+	} else if err != nil {
+		logging.FromContext(ctx).Error("Failed to get subscription", zap.Error(err))
+		// TODO: Send events here, or elsewhere?
+		//		r.Recorder.Eventf(p, corev1.EventTypeWarning, subscriptionCreateFailed, "Create Pipelines's subscription failed: %v", err)
```
same here, the error should be added to the log

vaikas-google replied:
I'm not sure I follow. What do you mean added to the log?

vaikas-google commented:
Thanks for the comments @n3wscott @lionelvillard please take another look. I also added code to properly set the address for the pipeline so it can be targeted, and also fixed the logic so that it tracks Channel CRs as they are created and fixed so that upon Subscription changes, pipeline gets reconciled.

lionelvillard commented:
@vaikas-google Long story short: it does not work for me. The reconciler is stuck on subscription.

my `pipeline.yaml`:

```
apiVersion: messaging.knative.dev/v1alpha1
kind: Pipeline
metadata:
  name: sample-pipeline
spec:
  steps:
  - apiVersion: serving.knative.dev/v1alpha1
    kind: Service
    name: pipeline-service1
  - apiVersion: serving.knative.dev/v1alpha1
    kind: Service
    name: pipeline-service2
  channelTemplate:
    apiVersion: eventing.knative.dev/v1alpha1
    kind: Channel
```

its status:
```
status:
  ChannelStatuses:
  - channel:
      apiVersion: eventing.knative.dev/v1alpha1
      kind: Channel
      name: sample-pipeline-kn-pipeline-0
    ready:
      lastTransitionTime: null
      status: ""
      type: ""
  - channel:
      apiVersion: eventing.knative.dev/v1alpha1
      kind: Channel
      name: sample-pipeline-kn-pipeline-1
    ready:
      lastTransitionTime: null
      status: ""
      type: ""
  SubscriptionStatuses: null
  address:
    hostname: sample-pipeline-kn-pipeline-0-channel-st4w4.default.svc.cluster.local
    url: http://sample-pipeline-kn-pipeline-0-channel-st4w4.default.svc.cluster.local
  conditions:
  - lastTransitionTime: "2019-06-05T15:09:44Z"
    status: "True"
    type: Addressable
  - lastTransitionTime: "2019-06-05T15:09:42Z"
    status: "True"
    type: ChannelsReady
  - lastTransitionTime: "2019-06-05T15:09:42Z"
    message: hostname is the empty string
    reason: emptyHostname
    status: "False"
    type: Ready
  - lastTransitionTime: "2019-06-05T15:09:42Z"
    status: Unknown
    type: SubscriptionsReady
```

And the controller logs:

```
knative-eventing eventing-controller-b75fbd969-q9nqj eventing-controller {"level":"info","ts":"2019-06-05T15:15:09.821Z","logger":"controller.pipeline-controller.event-broadcaster","caller":"record/event.go:221","msg":"Event(v1.ObjectReference{Kind:\"Pipeline\", Namespace:\"default\", Name:\"sample-pipeline\", UID:\"f22843d4-87a3-11e9-aeef-32340ea16cc7\", APIVersion:\"messaging.knative.dev/v1alpha1\", ResourceVersion:\"6836174\", FieldPath:\"\"}): type: 'Warning' reason: 'ReconcileFailed' Pipeline reconciliation failed: Internal error occurred: admission webhook \"webhook.eventing.knative.dev\" denied the request: mutation failed: missing field(s): spec.reply.channel.name","knative.dev/controller":"pipeline-controller"}
```

Missing the subscription channel name...


Here some notes I took while trying various configurations.

-  no error is logged in the CR when the version is missing in apiVersion (eg: `apiVersion: messaging.knative.dev`)
- When using InMemoryChannel CRD the conditions are wrong (see `ready` and `channel not ready`):

```
ChannelStatuses:
  - channel:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: InMemoryChannel
      name: sample-pipeline-kn-pipeline-0
    ready:
      lastTransitionTime: null
      status: ""
      type: ""
  - channel:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: InMemoryChannel
      name: sample-pipeline-kn-pipeline-1
    ready:
      lastTransitionTime: null
      status: ""
      type: ""
SubscriptionStatuses: null
  address:
    hostname: sample-pipeline-kn-pipeline-0-kn-channel.default.svc.cluster.local
    url: http://sample-pipeline-kn-pipeline-0-kn-channel.default.svc.cluster.local
  conditions:
  - lastTransitionTime: "2019-06-05T14:45:00Z"
    status: "True"
    type: Addressable
  - lastTransitionTime: "2019-06-05T14:45:00Z"
    status: "True"
    type: ChannelsReady
  - lastTransitionTime: "2019-06-05T14:45:00Z"
    message: Channels are not ready yet, or there are none
    reason: ChannelsNotReady
    status: "False"
    type: Ready
  - lastTransitionTime: "2019-06-05T14:45:00Z"
    status: Unknown
    type: SubscriptionsReady
```

Harwayne commented on config/200-controller-clusterrole.yaml:
```diff
@@ -76,6 +76,16 @@ rules:
+    resources:
+      - "pipelines"
+      - "pipelines/status"
+      - "inmemorychannels"
```
This will need to be an aggregated role. Similar to #1341, but include all permissions, not just read and update.

Harwayne commented on config/300-pipeline.yaml:
```diff
@@ -0,0 +1,91 @@
+    categories:
+    - all
+    - knative
+    - eventing
```
`- messaging` as well?

Harwayne commented on config/300-pipeline.yaml:
```diff
@@ -0,0 +1,91 @@
+            steps:
+              type: array
+              items:
+              properties:
```
I think this and everything below it needs to indent. As well as add `type: object`.

Harwayne commented on config/300-pipeline.yaml:
```diff
@@ -0,0 +1,91 @@
+                    - apiVersion
+                    - kind
+                    - name
+                  properties:
```
If you want, I think adding:

```yaml
namespace:
  type: string
  maxLength: 0
```

Might be useful.

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_lifecycle.go:
```diff
@@ -0,0 +1,165 @@
+			Subscription: corev1.ObjectReference{
+				APIVersion: s.APIVersion,
+				Kind:       s.Kind,
+				Name:       s.Name,
```
Namespace? (I'm not sure it is a great idea, but wanted to ask)

vaikas-google replied:
done. here and below.

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_types.go:
```diff
@@ -0,0 +1,154 @@
+	ChannelTemplate ChannelTemplateSpec `json:"channelTemplate"`
+
+	// Reply is a Reference to where the result of the last Subscriber gets sent to.
+	// Currently due to limitations on the Subscription, this _must_ be a Channel.
```
I think this is now out of date. @nachocano, I think you removed this requirement, right?

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+// reconcile the two. It then updates the Status block of the Pipeline resource
+// with the current Status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logging.FromContext(ctx).Error("reconciling", zap.String("key", key))
```
Either remove or set to debug.

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
```
, zap.Error(err)

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+		r.Recorder.Eventf(pipeline, corev1.EventTypeNormal, reconciled, "Pipeline reconciled")
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, pipeline.DeepCopy()); updateStatusErr != nil {
```
Why a second DeepCopy()? pipeline was already DeepCopied on line 131.

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+		return nil
+	}
+
+	//	obj := p.Spec.ChannelTemplate
```
Remove.

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+	if channelResourceInterface == nil {
+		msg := fmt.Sprintf("Unable to create dynamic client for: %+v", p.Spec.ChannelTemplate)
+		logging.FromContext(ctx).Error(msg)
+		errors.New(msg)
```
Is this meant to be returned?

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+	// Tell tracker to reconcile this Pipeline whenever my channels change.
+	track := r.addressableInformer.TrackInNamespace(r.tracker, p)
+
+	channels := []*duckv1alpha1.Channelable{}
```
```suggestion
	channels := make([]*duckv1alpha1.Channelable, 0, len(p.Spec.Steps))
```

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+	}
+	p.Status.PropagateChannelStatuses(channels)
+
+	subs := []*eventingv1alpha1.Subscription{}
```
```suggestion
	subs := make([]*eventingv1alpha1.Subscription, 0, len(p.Spec.Steps))
```

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+				logging.FromContext(ctx).Error(fmt.Sprintf("Failed to create Channel: %s/%s", p.Namespace, channelName), zap.Error(err))
+				return nil, err
+			}
+			logging.FromContext(ctx).Error(fmt.Sprintf("Created Channel: %s/%s", p.Namespace, channelName), zap.Any("NewChannel", zap.Any("NewChannel", newChannel)))
```
```suggestion
			logging.FromContext(ctx).Error(fmt.Sprintf("Created Channel: %s/%s", p.Namespace, channelName), zap.Any("NewChannel", newChannel))
```

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+			return nil, err
+		}
+	}
+	logging.FromContext(ctx).Error(fmt.Sprintf("Found Channel: %s/%s", p.Namespace, channelName), zap.Any("NewChannel", c))
```
Remove or debug level.

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+			return err
+		}
+		subs = append(subs, sub)
+		logging.FromContext(ctx).Info(fmt.Sprintf("Reconciled Subscription Object for step: %d: %+v", i, sub))
```
```suggestion
		logging.FromContext(ctx).Debug(fmt.Sprintf("Reconciled Subscription Object for step: %d: %+v", i, sub))
```

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+	// If the resource doesn't exist, we'll create it.
+	if apierrs.IsNotFound(err) {
+		sub = expected
+		logging.FromContext(ctx).Info(fmt.Sprintf("Creating subscription: %+v : SUBSCRIBERSPEC: %+v", sub, sub.Spec.Subscriber))
```
Isn't SUBSCRIBERSPEC already in subscription?

Harwayne commented on pkg/reconciler/pipeline/pipeline.go:
```diff
@@ -0,0 +1,297 @@
+	return reconcileErr
+}
+
+func pipelineSubscriptionName(pipelineName string, step int) string {
```
Redundant with resources/subscription.go's `PipelineSubscriptionName`. Remove one of them (my preference is to remove this one).

vaikas-google replied:
yeah, I moved it there and changed refs to it, but forgot to remove this from here, removed.

vaikas-google commented:
I wired a pipeline like this (using IMC (After my fix that got merged earlier today)), and the following pipeline worked, verified the data flew through the pipeline:
apiVersion: messaging.knative.dev/v1alpha1
kind: Pipeline
metadata:
  name: vaikas-pipeline
  namespace: foobar
spec:
  channelTemplate:
    apiVersion: messaging.knative.dev/v1alpha1
    kind: InMemoryChannel
  steps:
  - ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: first
  - ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: second
  - ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: third

vaikas-google commented:
If it looks good, I'll work on the validation / docs after this, but given the shift in time, unless there's something major that can't be fixed in a followup, I'd like to get this merged so that I don't have to keep rebasing :)
@lionelvillard @Harwayne

n3wscott commented:
/lgtm

Harwayne commented:
/hold

I want to review the changes.

Harwayne commented on pkg/apis/messaging/v1alpha1/pipeline_lifecycle.go:
```diff
@@ -100,25 +105,32 @@ func (ps *PipelineStatus) PropagateChannelStatuses(channels []*duckv1alpha1.Chan
-		if address == nil {
+		address := c.Status.AddressStatus.Address
+		if address != nil {
+			ps.ChannelStatuses[i].ReadyCondition = apis.Condition{Type: apis.ConditionReady, Status: corev1.ConditionTrue}
```
This isn't quite accurate, a Channelable may be addressable but not ready. 

For now, I'm OK with a TODO.

vaikas-google replied:
Yes, I had a TODO there (and dropped it somewhere along the way), and I think we need to change Channelable a little bit, because it doesn't include anything useful besides this.
